src/ui/features/turnstile/Turnstile.tsx

```diff
@@ -0,0 +1,62 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { useWindowSizeStore } from 'src/ui/shared/useWindowSizeStore';
+import { invariant } from 'src/shared/invariant';
+import { emitter } from 'src/ui/shared/events';
+import { isObj } from 'src/shared/isObj';
+import type { HTMLDialogElementInterface } from 'src/ui/ui-kit/ModalDialogs/HTMLDialogElementInterface';
+import { BottomSheetDialog } from 'src/ui/ui-kit/ModalDialogs/BottomSheetDialog';
+import { queryClient } from 'src/ui/shared/requests/queryClient';
+import { WebAppMessageHandler } from '../referral-program/WebAppMessageHandler';
+
+function TurnstileDialog() {
+  const { innerWidth } = useWindowSizeStore();
+  const turnstileWidgetHeight = 88;
+  const turnstileWidgetWidth = innerWidth - 32;
+  const dialogRef = useRef<HTMLDialogElementInterface | null>(null);
+
+  useEffect(() => {
+    dialogRef.current?.showModal();
+  }, []);
+
+  const handleTurnstileToken = useCallback((params: unknown) => {
+    invariant(
+      isObj(params) && typeof params.token === 'string',
+      'Got invalid payload from set-turnstile-token web app message'
+    );
+    dialogRef.current?.close();
+    queryClient.refetchQueries();
+  }, []);
+
+  return (
+    <BottomSheetDialog
+      ref={dialogRef}
+      height="fit-content"
+      closeOnClickOutside={false}
+      containerStyle={{ backgroundColor: 'var(--z-index-0)' }}
+    >
+      <WebAppMessageHandler
+        pathname="/turnstile.html"
+        callbackName="set-turnstile-token"
+        callbackFn={handleTurnstileToken}
+        hidden={false}
+        style={{
+          width: turnstileWidgetWidth,
+          height: turnstileWidgetHeight,
+          border: 'none',
+          colorScheme: 'auto',
+        }}
+      />
+    </BottomSheetDialog>
+  );
+}
+
+export function TurnstileTokenHandler() {
+  const [showDialog, setShowDialog] = useState(false);
+  useEffect(() => {
+    return emitter.on('openTurnstile', () => {
+      setShowDialog(true);
```

Comment:
Hm, do we never set this back to `false`? I guess it's minor but it feels incomplete:)